src/backend/distributed/operations/worker_split_shard_replication_setup_udf.c

```diff
@@ -0,0 +1,448 @@
+ */
+#include "postgres.h"
+#include "miscadmin.h"
+#include "postmaster/postmaster.h"
+#include "common/hashfn.h"
+#include "distributed/shardinterval_utils.h"
+#include "distributed/shard_utils.h"
+#include "distributed/shardsplit_shared_memory.h"
+#include "distributed/connection_management.h"
+#include "distributed/citus_safe_lib.h"
+#include "distributed/listutils.h"
+#include "distributed/remote_commands.h"
+#include "distributed/tuplestore.h"
+#include "distributed/shardsplit_logical_replication.h"
+#include "utils/builtins.h"
+#include "utils/lsyscache.h"
+#include "commands/dbcommands.h"
+
+
+/* declarations for dynamic loading */
+PG_FUNCTION_INFO_V1(worker_split_shard_replication_setup);
+
+static HTAB *ShardInfoHashMap = NULL;
+
+/* Function declarations */
+static void ParseShardSplitInfoFromDatum(Datum shardSplitInfoDatum,
+										 uint64 *sourceShardId,
+										 uint64 *childShardId,
+										 int32 *minValue,
+										 int32 *maxValue,
+										 int32 *nodeId);
+
+static ShardSplitInfo * CreateShardSplitInfo(uint64 sourceShardIdToSplit,
+											 uint64 desSplitChildShardId,
+											 int32 minValue,
+											 int32 maxValue,
+											 int32 nodeId);
+static void AddShardSplitInfoEntryForNodeInMap(ShardSplitInfo *shardSplitInfo);
+static void PopulateShardSplitInfoInSM(ShardSplitInfoSMHeader *shardSplitInfoSMHeader,
+									   HTAB *shardInfoHashMap);
+
+static void ReturnReplicationSlotInfo(HTAB *shardInfoHashMap, Tuplestorestate *tupleStore,
+									  TupleDesc tupleDescriptor);
+
+/*
+ * worker_split_shard_replication_setup UDF creates in-memory data structures
+ * to store the meta information about the shard undergoing split and new split
+ * children along with their placements. This info is required during the catch up
+ * phase of logical replication.
+ * This meta information is stored in a shared memory segment and accessed
+ * by logical decoding plugin.
+ *
+ * Split information is given by user as an Array of custom data type 'citus.split_shard_info'.
+ * (worker_split_shard_replication_setup(citus.split_shard_info[]))
+ *
+ * Fields of custom data type 'citus.split_shard_info':
+ * source_shard_id - id of the shard that is undergoing a split
+ *
+ * child_shard_id  - id of shard that stores a specific range of values
+ *                   belonging to sourceShardId(parent)
+ *
+ * shard_min_value - lower bound(inclusive) of hash value which childShard stores.
+ *
+ * shard_max_value - upper bound(inclusive) of hash value which childShard stores
+ *
+ * node_id         - Node where the childShardId is located
+ *
+ * The function parses the data and builds routing map with key for each distinct
+ * <nodeId, tableOwner> pair. Multiple shards can be placed on the same destination node.
+ * Source and destination nodes can be same too.
+ *
+ * There is a 1-1 mapping between a table owner and a replication slot. One replication
```

Comment:
Should we rephrase as : There is a 1-1 mapping between a (table owner, node) and replication slot.